pkgs/development/tools/continuous-integration/github-runner/default.nix

```diff
@@ -224,21 +224,35 @@ stdenv.mkDerivation rec {
 
     # Install the helper scripts to bin/ to resemble the upstream package
     mkdir -p $out/bin
-    install -m755 src/Misc/layoutbin/runsvc.sh        $out/bin/
-    install -m755 src/Misc/layoutbin/RunnerService.js $out/lib/
-    install -m755 src/Misc/layoutroot/run.sh          $out/lib/
-    install -m755 src/Misc/layoutroot/config.sh       $out/lib/
-    install -m755 src/Misc/layoutroot/env.sh          $out/lib/
+    install -m755 src/Misc/layoutbin/runsvc.sh                 $out/bin/
+    install -m755 src/Misc/layoutbin/RunnerService.js          $out/lib/
+    install -m755 src/Misc/layoutroot/run.sh                   $out/lib/
+    install -m755 src/Misc/layoutroot/run-helper.sh.template   $out/lib/run-helper.sh
+    install -m755 src/Misc/layoutroot/config.sh                $out/lib/
+    install -m755 src/Misc/layoutroot/env.sh                   $out/lib/
 
     # Rewrite reference in helper scripts from bin/ to lib/
-    substituteInPlace $out/lib/run.sh    --replace '"$DIR"/bin' "$out/lib"
-    substituteInPlace $out/lib/config.sh --replace './bin' "$out/lib"
+    substituteInPlace $out/lib/run.sh    --replace '"$DIR"/bin' '"$DIR"/lib'
+    substituteInPlace $out/lib/config.sh --replace './bin' $out'/lib'
+
+    # Remove uneeded copy for run-helper template
+    substituteInPlace $out/lib/run.sh --replace 'cp -f "$DIR"/run-helper.sh.template "$DIR"/run-helper.sh' ' '
+    substituteInPlace $out/lib/run-helper.sh --replace '"$DIR"/bin/' '"$DIR"/'
 
     # Make paths absolute
     substituteInPlace $out/bin/runsvc.sh \
       --replace './externals' "$out/externals" \
       --replace './bin' "$out/lib"
 
+    # Set working dir in config.sh
+    sed -i '2 i cd $RUNNER_ROOT' $out/lib/env.sh
+    sed -i '2 i cd '$out/ $out/lib/config.sh
+
+    # Set default value for RUNNER_ROOT
+    sed -i '2 i export RUNNER_ROOT="$\{RUNNER_ROOT:-$HOME/.github-runner}"' $out/lib/config.sh
+    sed -i '3 i mkdir -p $RUNNER_ROOT' $out/lib/config.sh
+    sed -i '2 i export RUNNER_ROOT="$\{RUNNER_ROOT:-$HOME/.github-runner}"' $out/lib/run.sh
+
```

Comment:
I'd try to integrate those changes into `host-context-dirs.patch` which introduces the `RUNNER_ROOT` environment variable.

Reply:
I agree that could be cleaner. What should we do about the root dir in `env.sh` then? It's called in `config.sh` before the dotnet package in invoked at all. We'd need to hardcode the runner root default in both places, which actually feels worse. For reference: https://github.com/actions/runner/blob/main/src/Misc/layoutroot/env.sh

Reply:
Also, I think these specific calls to `sed` are actually quite resilient. `config.sh` and `run.sh` are essentially the public APIs of the runner package, and placing these patches in the second line of the script doesn't assume much except that it's a bash session.

<img width="328" alt="Screen Shot 2022-08-15 at 10 35 58 AM" src="https://user-images.githubusercontent.com/7585078/184655802-4a66caee-70af-413f-acc7-0ed06407ecbd.png">

Reply:
So far, we created wrappers in `bin/` for the shell scripts. As far as I can tell, [`makeWrapper`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/make-wrapper.sh) (particularly, `--set-default` and `--chdir`) would work well here as well (though I'd move the invocations to `postFixup`).

Reply:
Thanks for the pointers. I've refactored the change to use those helper functions. Looks much cleaner now.

Given some quirks with bash quote escaping, I wasn't able to use `--set-default`, but `--run` did the trick.